### ISSUE TITLE
NVCC Warning under Visual Studio

### ISSUE THREAD

**Issue by Robadob** [FLAMEGPU/FLAMEGPU2_dev#106]
Waiting on CMake feature to be added: [view progress here](https://gitlab.kitware.com/cmake/cmake/issues/18265)

------------
When building the examples under Visual Studio with warning level set to `/W4`, an nvcc internal process causes the below warning to be produced many times over.

e.g. `C:/Users/appveyor/AppData/Local/Temp/tmpxft_00000698_00000000-6_simple_example.device-link.reg.c(17): warning C4100: 'prelinked_fatbinc': unreferenced formal parameter [C:\projects\FLAMEGPU2_dev\build\examples\simple_example\simple_example.vcxproj]`

For some reason, promoting warnings to error's doesn't cause this to error, so it's not that big of a deal, but I'd rather fix it.

unreferenced formal parameter is a useful warning to have enabled, as it can identify when a typo or similar has caused a parameter to not be used. The 'official' way to leave intentionally redundant parameters in-place is to comment out their identifier.

[This](https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html) NVidia compiler guide has the below diagram, in which `a_dlink.reg.c` appears to be the file reporting the warning. This confirms the warning is coming from NVCCs link process.
![image](https://user-images.githubusercontent.com/742154/68054288-db643a80-fce5-11e9-8eaa-bddd4609e617.png)

**Comment by Robadob** [comment]
Done some testing.

We should add to CUDA C/C++::additional_options:
```
--Wreorder --Werror reorder,cross-execution-space-call -Xptxas="-Werror"  -Xnvlink="-Werror"
```
We should add to CUDA C/C++::additional_options:
```
-Xcompiler="/wd4100 /WX"
```

**Comment by Robadob** [comment]
I've applied the changes, however `CUDA_DEVICE_LINK_FLAGS` is not yet supported, so the warnings persist.

Nudged a CMake repo post from 3 months ago about it: https://gitlab.kitware.com/cmake/cmake/issues/18265

**Comment by Robadob** [comment]
Appears it had been forgotten, so the nudge might help get it fixed in a future release.

![image](https://user-images.githubusercontent.com/742154/68074503-8e3ea200-fd93-11e9-892b-359c12a62a7f.png)

**Comment by Robadob** [comment]
The CMake fix has been merged as of today: https://gitlab.kitware.com/cmake/cmake/-/merge_requests/4577

I think the fix will be to pass the link option `$<DEVICE_LINK:-Xcompiler="/wd4100 /WX">`

Maybe also need to check that CMake version is greater than 3.17.1

Will need to remove the previous assumed fix with `CUDA_DEVICE_LINK_FLAGS`.

**Comment by Robadob** [comment]
Cmake 3.17.3 is now available (and 3.18 imminent), this means this can now be resolved.
